tes-adapter/src/main/java/com/epam/pipeline/tesadapter/service/TaskMapper.java

```diff
@@ -65,8 +80,54 @@ public PipelineStart mapToPipelineStart(TesTask tesTask) {
     }
 
     private TesExecutor getExecutorFromTesExecutorsList(List<TesExecutor> tesExecutors) {
-        Assert.notEmpty(tesExecutors, messageHelper.getMessage(MessageConstants.ERROR_PARAMETER_NULL_OR_EMPTY,
-                "executors"));
+        Assert.isTrue(tesExecutors.size() == ONLY_ONE, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_INCOMPATIBLE_CONTENT, EXECUTORS, tesExecutors));
         return tesExecutors.get(FIRST);
     }
+
+    private String getProperInstanceType(TesTask tesTask, Tool pipelineTool) {
+        Double ramGb = tesTask.getResources().getRamGb();
+        Long cpuCores = tesTask.getResources().getCpuCores();
+        Long toolId = pipelineTool.getId();
+        Long regionId = getProperRegionIdInCloudRegionsByTesZone(tesTask.getResources().getZones());
+        Boolean spot = tesTask.getResources().getPreemptible();
+
+        AllowedInstanceAndPriceTypes allowedInstanceAndPriceTypes = cloudPipelineAPIClient
+                .loadAllowedInstanceAndPriceTypes(toolId, regionId, spot);
+        Assert.isTrue(allowedInstanceAndPriceTypes.getAllowedInstanceTypes() != null, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_NULL_OR_EMPTY, allowedInstanceAndPriceTypes));
+        return evaluateMostProperInstanceType(allowedInstanceAndPriceTypes, ramGb, cpuCores);
+    }
+
+    private Tool loadToolByTesImage(String image) {
+        return cloudPipelineAPIClient.loadTool(image);
+    }
+
+    private Long getProperRegionIdInCloudRegionsByTesZone(List<String> zones) {
+        Assert.isTrue(zones.size() == ONLY_ONE, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_INCOMPATIBLE_CONTENT, ZONES, zones));
+        return cloudPipelineAPIClient.loadAllRegions().stream().filter(
+                region -> region.getName().equalsIgnoreCase(zones.get(FIRST)))
+                .collect(Collectors.toList()).get(FIRST).getId();
```

Comment:
will it work if there is no regions that pass filter?

Reply:
@evgeniimv So, was this problem solved?

Reply:
Solved. Also added the same check in **evaluateMostProperInstanceType()** method for **allowedInstanceAndPriceTypes.getAllowedInstanceTypes()** method. Check that not-null (**Optional.ofNullable()**)